### ISSUE TITLE
stream.dash: RuntimeError: cannot schedule new futures after interpreter shutdown

### ISSUE THREAD

**Issue by BluchipStudio** [streamlink/streamlink#4626]
### Checklist

- [X] This is a bug report and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed bug reports](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22bug%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest stable release

### Description

I am trying to run streamlink under ubuntu 20 where python 3.8 comes preinstalled by default

when i try and start a dash stream i get the following error this error does not happen if i use python 7 with ubuntu 18

### Debug log

```text
00:02:54,927 server.7860b DEBUG Exception in thread Thread-DASHStreamWorker:
00:02:55,439 server.7860b DEBUG Traceback (most recent call last):
00:02:55,440 server.7860b DEBUG   File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
00:02:55,440 server.7860b DEBUG     self.run()
00:02:55,440 server.7860b DEBUG   File "/usr/local/lib/python3.8/dist-packages/streamlink/stream/segmented.py", line 87, in run
00:02:55,440 server.7860b DEBUG     self.writer.put(segment)
00:02:55,440 server.7860b DEBUG   File "/usr/local/lib/python3.8/dist-packages/streamlink/stream/segmented.py", line 143, in put
00:02:55,440 server.7860b DEBUG     future = self.executor.submit(self.fetch, segment, retries=self.retries)
00:02:55,440 server.7860b DEBUG   File "/usr/lib/python3.8/concurrent/futures/thread.py", line 181, in submit
00:02:55,440 server.7860b DEBUG     raise RuntimeError('cannot schedule new futures after '
00:02:55,440 server.7860b DEBUG RuntimeError: cannot schedule new futures after interpreter shutdown
00:02:55,440 server.7860b DEBUG Exception in thread Thread-DASHStreamWorker:
00:03:39,275 server.7860b DEBUG Traceback (most recent call last):
```

**Comment by back-to** [comment]
run streamlink from your terminal

`streamlink DASH-URL best -l debug` the output will be more helpful

---

if the stream works on ubuntu 18
`ffmpeg` might be missing

try

```
ffmpeg -v
```

or install it

```
sudo apt update
sudo apt install ffmpeg
```

**Comment by bastimeyer** [comment]
The reason why the `RuntimeError` exception gets raised is because the `DASHStreamWorker` thread is trying to add a new segment to the `DASHStreamWriter` after the main thread was already terminated.

This can either be due to a missing `stream.close()` call, which would prevent the worker thread from doing that, or because of a race condition in the `DASHStreamWorker` where it's using `sleep()` calls rather than `threading.Event`s for waiting on new segments (#4311).

As said, the `RuntimeError` should not be deterministic, and python 3.7 or python 3.8 shouldn't matter here either, unless there was a change in Python 3.8 which made the thread pool executor raise the `RuntimeError` which I'm not aware of (haven't looked that up). Either way, the DASHStream implementation needs to get fixed.

I have a local fix for the `sleep()` calls since a couple of months which I haven't pushed and opened a PR for yet, because there are two separate `sleep()` calls, and I only have fixed one, namely the one in the [`DASHStreamWorker.iter_segments()`](https://github.com/streamlink/streamlink/blob/2c564dbe18a4d18ed048a2e832655617a01a508a/src/streamlink/stream/dash.py#L89) method, which is relatively easy to fix, as there's already the `self._wait` `threading.Event` object that can be used, similar to the `HLSStreamWriter`. The one in [`DASHStreamWriter.fetch()`](https://github.com/streamlink/streamlink/blob/2c564dbe18a4d18ed048a2e832655617a01a508a/src/streamlink/stream/dash.py#L34) however requires a bit of refactoring, as the `fetch` method is executed by each thread in the writer's thread pool.

----

I'm re-labelling this as a bug, since the `RuntimeError` should not be raised here. But as @back-to already said, FFmpeg is a requirement for DASH streams, and this might be the culprit for your playback issues.

Since the debug log has been truncated, which you were not supposed to, there's not much more to debug here.
**Please provide the FULL DEBUG LOG**, as requested by the issue template. DON'T IGNORE IT.

**Comment by bastimeyer** [comment]
This was fixed by #4630. The issue here wasn't linked by mistake.
